src/Core/L10n.php

```diff
@@ -0,0 +1,243 @@
+		$lang = $language;
+	}
+
+	/**
+	 * Pop language off the top of the stack
+	 */
+	public static function popLang()
+	{
+		global $lang, $a;
+
+		if ($lang === $a->langsave) {
+			return;
+		}
+
+		if (isset($a->stringsave)) {
+			$a->strings = $a->stringsave;
+		} else {
+			$a->strings = [];
+		}
+
+		$lang = $a->langsave;
+	}
+
+	/**
+	 * load string translation table for alternate language
+	 *
+	 * first addon strings are loaded, then globals
+	 *
+	 * @param string $lang language code to load
+	 */
+	public static function loadTranslationTable($lang)
+	{
+		$a = get_app();
+
+		$a->strings = [];
+		// load enabled addons strings
+		$addons = dba::select('addon', ['name'], ['installed' => true]);
+		while ($p = dba::fetch($addons)) {
+			$name = $p['name'];
+			if (file_exists("addon/$name/lang/$lang/strings.php")) {
+				include("addon/$name/lang/$lang/strings.php");
+			}
+		}
+
+		if (file_exists("view/lang/$lang/strings.php")) {
+			include("view/lang/$lang/strings.php");
+		}
+	}
+
+	/**
+	 * @brief Return the localized version of the provided string with optional string interpolation
+	 *
+	 * This function takes a english string as parameter, and if a localized version
+	 * exists for the current language, substitutes it before performing an eventual
+	 * string interpolation (sprintf) with additional optional arguments.
+	 *
+	 * Usages:
+	 * - t('This is an example')
+	 * - t('URL %s returned no result', $url)
+	 * - t('Current version: %s, new version: %s', $current_version, $new_version)
+	 *
+	 * @param string $s
+	 * @return string
+	 */
+	public static function t($s)
+	{
+		$a = get_app();
+
+		if (x($a->strings, $s)) {
+			$t = $a->strings[$s];
+			$s = is_array($t) ? $t[0] : $t;
+		}
+		if (func_num_args() > 1) {
+			$args = array_slice(func_get_args(), 1);
+			$s = @vsprintf($s, $args);
+		}
+
+		return $s;
+	}
+
+	/**
+	 * @brief Return the localized version of a singular/plural string with optional string interpolation
+	 *
+	 * This function takes two english strings as parameters, singular and plural, as
+	 * well as a count. If a localized version exists for the current language, they
+	 * are used instead. Discrimination between singular and plural is done using the
+	 * localized function if any or the default one. Finally, a string interpolation
+	 * is performed using the count as parameter.
+	 *
+	 * Usages:
+	 * - tt('Like', 'Likes', $count)
+	 * - tt("%s user deleted", "%s users deleted", count($users))
+	 *
+	 * @global type $lang
+	 * @param string $singular
+	 * @param string $plural
+	 * @param int $count
+	 * @return string
+	 */
+	public static function tt($singular, $plural, $count)
+	{
+		global $lang;
+		$a = get_app();
+
+		if (x($a->strings, $singular)) {
+			$t = $a->strings[$singular];
+			if (is_array($t)) {
+				$plural_function = 'string_plural_select_' . str_replace('-', '_', $lang);
+				if (function_exists($plural_function)) {
+					$plural_function = 'stringPluralSelectDefault';
```

Comment:
This line is not working. I had it as 'self::stringPluralSelectDefault' and that did not work either. I just saw an error message for it this morning. I guess it means I don't know how to set a variable to a private static class function?

Reply:
`self::$plural_function()`

Reply:
So, in this instance
`If (function_exists($plural_function)) {
    $plural_function = self::$stringPluralSelectDefault();`

Reply:
````
$plural_function = 'string_plural_select_' . str_replace('-', '_', $lang);
if (function_exists($plural_function)) {
	$plural_function = 'stringPluralSelectDefault';
}
 $i = self::$plural_function($count);
````

Reply:
Actually no, the plural functions are automatically generated by `util/po2php.php` and are global, so this wouldn't work, here's the actual code:
````php
$plural_function = 'string_plural_select_' . str_replace('-', '_', $lang);
if (function_exists($plural_function)) {
	$i = $plural_function($count);
} else {
	$i = self::stringPluralSelectDefault($count);
}
````